### ISSUE TITLE
Write a Bitmap Font Importer

### ISSUE THREAD

**Issue by ilexp** [AdamsLair/duality#286]
The current way of importing Fonts into Duality is working, but quite limited and a bit buggy at times due to less-than-ideal Font handling in GDI+ / System.Drawing. A viable alternative would potentially improve usability and at the same time provide a sample on how to write custom Bitmap Font Importers.

Pick a commonplace bitmap font generator and implement a custom importer for it. Specifically, the [AngelCode Bitmap Font Generator](http://www.angelcode.com/products/bmfont/) looks quite good.

This can be done easily in a distinct plugin and no special core code is necessary for this.

**Comment by mika76** [comment]
Hey @ilexp I'd love to try help out with this, and it sounds pretty simple, I'm just wondering how the actual importing would work from a user point of view - since there are two files involved, a BMFont file and the actual image. Is it assumed they are named the same and that a user would drag in the BMFont file?

Also are there any similar custom importers I could use as a guideline?

**Comment by ilexp** [comment]
With the new Asset import / export API of v2.0, an import operation can involve multiple files and you can explicitly specify that N different files make up one single Resource. 

Ideally, the user would dragdrop all involved files into the editor at once, but it is up to your importer implementation whether missing files are tracked down based on path assumptions such as being in the same directory and named all the same. You could also scan the BMFont file for filename cues in that case, should the format specify them.

> Also are there any similar custom importers I could use as a guideline?

Unfortunately not - all existing importers have a 1:1 mapping between imported files and generated resources. However, I tried to supply all the required information via code documentation. As a general overview, you can take a look at the [IAssetImporter](https://github.com/AdamsLair/duality/blob/master/DualityEditor/AssetManagement/IAssetImporter.cs) and [IAssetImportEnvironment](https://github.com/AdamsLair/duality/blob/master/DualityEditor/AssetManagement/Import/IAssetImportEnvironment.cs) interface definitions, as well as existing [importer implementations](https://github.com/AdamsLair/duality/tree/master/DualityEditorPlugins/EditorBase/Importers) and the [extension methods](https://github.com/AdamsLair/duality/blob/master/DualityEditor/AssetManagement/Import/ExtMethodsIAssetImportEnvironment.cs) they use to share and simplify code. You will likely not use the extension methods since they're targeting 1:1 mapping scenarios, but they're documented to help you understand what's going on.

**Comment by mika76** [comment]
@ilexp Is there any way to stop an import of a file - currently if I drag drop in both a .fnt and a .png file, I can create the font off the .fnt file, but the png is picked up by the SpriteImporter as well and is imported as a Pixmap.

Also, is export required, and in this case what would be expected?

**Comment by mika76** [comment]
@ilexp It's going great so far, but I'm struggling a little to align the characters to the baseline. The BMFont definition has a yoffset value per char but the Glyph class does not. How do you calculate the y position for each char? I also don't know what the difference is between Ascent and BodyAscent.

I understand that kernings are offsets to a character when it is placed next to another specific character but it does not seem like the current Font implementation uses it like this? I do not understand the fat that it's just a int[] on each side, but in some cases you check there are 6 segments? Could you explain this a little or push me in the tight direction to some information?

I have been reading and re-reading the Font class but I'm a little cross-eyed :wink:  by now...

![2016-04-16 2](https://cloud.githubusercontent.com/assets/229311/14579852/6511e5cc-03b9-11e6-909f-9a42cca2cde5.png)

**Comment by ilexp** [comment]
@mika76 Nice work so far! :) I'll just quickly go over some of your questions:

> Is there any way to stop an import of a file - currently if I drag drop in both a .fnt and a .png file, I can create the font off the .fnt file, but the png is picked up by the SpriteImporter as well and is imported as a Pixmap.

The key here is not to stop other import operations, but to tell the system your importer is going to handle the .png file as well. In the `PrepareImport` step, make sure to call `env.HandleInput` on all files you're going to use. That way, the import system will be able to detect the conflict and spawn a user dialog which importer to prefer. To modify default behavior here, you can also return a higher-than-zero value in the `Priority` property, since your importers use case (font AND image) is more specialized than the Pixmap import case (just an image).

> Also, is export required, and in this case what would be expected?

No, export is not required. It serves two purposes:

1. Determining which source files belong to the imported Resource. This is used to decide what should happen when double-clicking the Resource, as well as to synchronize the `Source/Media` directory with moved or renamed files in `Data`.
2. Extracting previously imported data for editing.

You can safely omit export functionality entirely, it's an optional feature - though users might expect it to be there. Double-clicking a bitmap font could easily be a convenient "Okay, let me edit the image file of the font by hand now" - that's one thing the export would allow.

> It's going great so far, but I'm struggling a little to align the characters to the baseline. The BMFont definition has a yoffset value per char but the Glyph class does not. How do you calculate the y position for each char?

Duality fonts use glyphs that have a constant height, so that offset needs to be encoded in the image. However, I do recognize that this is kind of a nonsense requirement that merely arises for historical reasons. 

If you're ready to take the extra mile and dive into the Duality part of this, you can add an `OffsetY` field to `GlyphData` and implement the Y offset logic in [ProcessTextAdv](https://github.com/AdamsLair/duality/blob/master/Duality/Resources/Font.cs#L879) and all the methods using that one. As long as you don't modify any public API except for the added `GlyphData` field, this should not be a breaking change and could be released as a regular update. Feel free to do a Pull Request on this, should you go down that road.

The other alternative would be to simply adjust the pixel data after loading, but before passing it to the `Font` resource. This can be a part of the import process and does not require any core changes, thus you won't have to go through the core dev / Pull Request / review / approval / release cycle stuff. 

> I also don't know what the difference is between Ascent and BodyAscent.

See this Wikipedia image for reference:
![Typogrphy Line Terms](https://upload.wikimedia.org/wikipedia/commons/thumb/3/39/Typography_Line_Terms.svg/2000px-Typography_Line_Terms.svg.png)
While the Ascent value represents the "ascender height" from the image, BodyAscent refers to the "x-height", or height of the median line.

> I understand that kernings are offsets to a character when it is placed next to another specific character but it does not seem like the current Font implementation uses it like this? I do not understand the fat that it's just a int[] on each side, but in some cases you check there are 6 segments? Could you explain this a little or push me in the tight direction to some information?

Duality doesn't use kerning pairs, mostly for historic reasons, since the default System.Drawing.Font based importer simply has no way of accessing this data and thus cannot provide it. Instead, fonts in Duality do kerning based on sampling how many transparent pixels there are at the side of each glyph using fixed "sample heights". 
Each of the `int[]` indices represents one height, and the number in the array represents the number of transparent pixels from the respective side. All of this is calculated automatically and there's not really a way to use a different kerning algorithm right now. 

This isn't ideal and will certainly change at some point, but unlike the `OffsetY` issue I wouldn't recommend you to tackle this in a PR since that part of code might be subject to a later (v3.0?) redesign and I'd like to give this some thought. So for now, I think you could just discard the provided kerning pairs data and hope for the best.

So.. that turned out to be a bigger answer than anticipated. Hope I could help you clear some things up!

**Comment by mika76** [comment]
Thanks @ilexp for the detailed answer :smile: 

I went with the route of changing the Font class to allow for a offsetX field - works really well. Currently the importer works quite well with some limitations:
* I am only parsing the xml version o BMFont. I'm not sure how much the text file version is required. XML has the same data all the programs I've tried so far export it, so not sure if the effort of writing the parser is warranted. Time will tell...
* It seem the pixeldata class draws a rectangle when there is an atlas involved? Or is this just in the visulalizer? I'm finding if I dont include a padding when generating the font, I'm seeing some extra lines - I'm not sure if this is from the pixeldata or from the calculations used int the front placing, but that's tomorrow's problem :smile: 
* BMFont saves it's font data in specific channels (no colours or transparency), which I'm not dealing with currently (it was done this way I guess to allow for using the font data with a material or something). Currently they are displayed with the black background since I don't manipulate the texture currently.
* BMFont format allows for multiple pages as attachments, but currently only a single page is supported. I'm not sure if I will support more since it means I would have to create a custom bitmap and re-calculate the glyps considering the Font class does not support it. Not sure it's worth the trouble.

Any comments on the above?

![2016-04-17](https://cloud.githubusercontent.com/assets/229311/14584096/e4cdd9b0-0436-11e6-8833-92bdda11b46c.png)

**Comment by mika76** [comment]
Ps: Do you have any helper classes I can use to extract a single channel like alpha onto it's own bitmap or something like that?

**Comment by mika76** [comment]
Pps: Knowing the code a little better now, I'm thinking I could add kerning to the Font class pretty easily - are you sure you'd like me to stay away from that? It seems they can be extracted from a font using PInvoke method http://stackoverflow.com/questions/16699862/how-do-you-use-windows-gdi-methods-in-net-that-arent-in-gdi so even the normal font generator could use them.

**Comment by ilexp** [comment]
This is starting to look really nice. Thumbs up!

> It seem the pixeldata class draws a rectangle when there is an atlas involved? Or is this just in the visulalizer? I'm finding if I dont include a padding when generating the font, I'm seeing some extra lines - I'm not sure if this is from the pixeldata or from the calculations used int the front placing, but that's tomorrow's problem :smile: 

That's just the visualizer telling you there are atlas rectangles. I suspect the padding is required due to texture filtering, which is fine and kind of expected. The default font importer also uses padding, for that same reason. Disabling texture filtering on the other hand just doesn't look nice when scaling or rotating the font, so I think padding is a good solution here.

> Ps: Do you have any helper classes I can use to extract a single channel like alpha onto it's own bitmap or something like that?

Nope, no helper classes for this. You'll have to iterate over the pixeldata to do this.

> Pps: Knowing the code a little better now, I'm thinking I could add kerning to the Font class pretty easily - are you sure you'd like me to stay away from that? It seems they can be extracted from a font using PInvoke method http://stackoverflow.com/questions/16699862/how-do-you-use-windows-gdi-methods-in-net-that-arent-in-gdi so even the normal font generator could use them.

Introducing kerning pairs is a bit of a code design thing, which I'm always hesitant to assign to new contributors because it often takes a bit until someone really "gets into" the framework - and I really try to avoid handing out tasks and _then_ having to reject Pull Requests people have been putting serious work into. That said, you seem to be on a roll here, so if you're up to it, I'd say do it. Just two things I'd suggest:

1. Let's talk about how you're going to implement this before you start.
2. Do a separate Pull Request for the OffsetY thing and the kerning pairs, so I don't have to decide to take both or none. This means, you'll need one branch for each of them, but since the offset change is much more basic, I'd say it's fine if the kerning branch builds up on the offset branch.

Oh, and thanks for the stack overflow link! I wasn't aware this was actually possible.

> Any comments on the above?

Overall, I think it's a good thing you're starting with the basics and don't try to support every quirk and feature of the BMFont format right from the beginning. No need to go the extra mile for a feature if it turns out no one uses it.

Did you try the regular Duality kerning with BMFonts by the way? How did that hold up?

**Comment by mika76** [comment]
> Introducing kerning pairs...

Ok I'll finish up with the basics here then see about this.

> Overall, I think it's a good thing you're starting with the basics and don't try to support every quirk and feature of the BMFont format right from the beginning. No need to go the extra mile for a feature if it turns out no one uses it.

I agree. Also people were just not consistent. There are a couple of programs I have run into now which have totally different file formats.

> Did you try the regular Duality kerning with BMFonts by the way? How did that hold up?

No not yet - I will play with it a bit more and send through a pull request before moving onto other stuff...

**Comment by ilexp** [comment]
Allright, sounds good :+1:

**Comment by mika76** [comment]
@ilexp One question - in the editor, once I import a bitmap font, all the properties seem to be disabled. Do you maybe know why this is? I can't seem to trace it to anything specific.

**Comment by mika76** [comment]
@ilexp Found it - CanRenderFont is checking for an EmbeddedFont...

**Comment by ilexp** [comment]
I think you're looking for [this](https://github.com/AdamsLair/duality/blob/master/DualityEditorPlugins/EditorBase/PropertyEditors/Resources/FontPropertyEditor.cs#L40). It's in there because editing those properties only make sense if the editor can actually [re-render the font with those new settings](https://github.com/AdamsLair/duality/blob/master/DualityEditorPlugins/EditorBase/EditorBasePlugin.cs#L88), which is currently determined by checking whether the font includes an embedded TrueType font. This hardcoded kind of relation will [go away eventually](https://github.com/AdamsLair/duality/issues/285), but for now we'll have to deal with it.

(I'm kind of sorry for the rabbit hole you've fallen into with this issue - did not expect you to run into that amount of historic code / hacks :smile: )

Is this a problem with regard to your importer?

**Comment by mika76** [comment]
:smile: It's not a problem - best way to learn something is to dive right in I think. 

Also I fell into the trap of seeing the graphics render and thinking it's all working 100%, now I can see that changing the values is causing some issues I missed. One question - I can't think currently of any time you might have a Font _without_ an EmbeddedFont (currently, without the bitmap stuff)? So I'm playing with changing the logic slightly for those checks in the Font class. I will try out a bunch of scenarios and let you know how I get on. Also, soon I'll make a report so you can see what's been done. It probably sounds a lot bigger than it really is.

**Comment by ilexp** [comment]
> One question - I can't think currently of any time you might have a Font without an EmbeddedFont (currently, without the bitmap stuff)?

There's no need to keep it in there when releasing a game, even though the editor might be included in that release - so that's a case that should be supported, even though it doesn't happen in the "primary user story".

Also, as you noted, imported bitmap fonts - from your importer or others / custom ones - don't have an embedded truetype font, so there's that.

> Also, soon I'll make a report so you can see what's been done. It probably sounds a lot bigger than it really is.

Nice, looking forward to it.

**Comment by mika76** [comment]
> There's no need to keep it in there when releasing a game, even though the editor might be included in that release - so that's a case that should be supported, even though it doesn't happen in the "primary user story".

I thought by this you mean deleting the ttf file from the Media folder, but when I tried that, the EmbeddedFont still has it's full data (I tried with acustom, non-installed font). I'm assuming it's serialized? In which case how would it be cleared? 

Basically I'm trying to have a check whether a font is a bitmap font or an embedded font, without adding a new property :smile: the question is is it safe to assume that if EmbeddedFont is null it means this would be a bitmap font.

**Comment by ilexp** [comment]
> I thought by this you mean deleting the ttf file from the Media folder, but when I tried that, the EmbeddedFont still has it's full data (I tried with acustom, non-installed font). I'm assuming it's serialized? In which case how would it be cleared? 

It could be cleared either manually (deleting the property in the XML of the Font Resource) or as part of an automated deployment step (which nulls the property and re-saves the Font Resource). The `Media/Source` folder is not required or referred to by a Resource in any way, since they follow the overall goal of being self-contained within the Duality content system.

> Basically I'm trying to have a check whether a font is a bitmap font or an embedded font, without adding a new property :smile: the question is is it safe to assume that if EmbeddedFont is null it means this would be a bitmap font.

In fact, _all_ Duality Fonts are bitmap fonts! Some just happen to carry a `.ttf` around so the editor can re-render them when a user requests to do so. This feature, however, will be moved to the new Asset Importer data attachments (Issue #285), so the `EmbeddedFont` property will no longer exist in the long run. All that distinguishes Fonts would be the importer that is preferred for them, which is something that shouldn't be checked explicitly.

Is there a specific reason why you want to know what "kind" of Font you're dealing with?

**Comment by mika76** [comment]
Hahaha no prob. I'm expecting a good code thrashing. I respect that you take care of your code base :smile:

The reason I wanted to know was purely to not get the null checks wrong since the font has to be re-generated in some cases (eg size is changed)

But you'll see in the code...

**Comment by ilexp** [comment]
Allright, now that the Offset PR is all wrapped up and merged, what will be the next steps? Apart from the readonly issue that prevents users from toggling kerning, is there anything else that's missing for getting your plugin to work in a first iteration?

The readonly issue can be solved by some more specialized FontPropertyEditor code and will no longer exist after #285, but there's no ETA on that, so a workaround might be a good thing until then.

**Comment by mika76** [comment]
I was thinking about this. The main reason you make the properties readonly is because you need an embedded font to regenerate the glyphs. 

But for updating kerning you don't need to regenerate them. So how about excluding only the kerning property from becoming readonly. And on changing the kerning property the glyphs do not regenerate but the kerning arrays are recalculated.

**Comment by ilexp** [comment]
Yep, that's what I was trying to say with the second paragraph ^^ I can look into this, or you can do if you want to take this as a chance for diving into this.

Overall, what's the state / roadmap for your plugin?

**Comment by mika76** [comment]
Have a [look](https://github.com/mika76/duality-bitmapfont-importer) 😁 

You can try it by just adding the .cs file to the main Duality project for the moment - until I make an actual package.

**Comment by ilexp** [comment]
I have implemented a better readonly-switch in the FontPropertyEditor, so you can now switch kerning and various other properties on and off without an embedded .ttf Font. Let me know if that fixes the issue for your scenario.

> Have a look :grin:

Tried it. :+1: 

![bitmapfonttest](https://cloud.githubusercontent.com/assets/14859411/14910031/2d33ea20-0dea-11e6-9299-6dbcef55fd77.png)

Also, nice work on the info page!

That said, I did stumble across an issue - space characters appear to be really, really small. When writing a text, you can sometimes barely see the difference between two words, or just two characters of a word. I took a quick look at your [parsing code](https://github.com/mika76/duality-bitmapfont-importer/blob/master/DualityBitmapImporter/DualityBitmapImporter/BitmapFontImpoter.cs#L259) and it seems like you don't parse the `xadvance` attribute of a BMFont's glyphs, which should determine behavior in special cases where a char can be wider than its visual representation. 

When looking at a BMFont file (especially in one with outlines and shadow effects), you can see that in almost all cases, `xadvance` is different from `width`, so I think that this needs to be addressed. Unfortunately, Duality Fonts do not distinguish between width and x-advancement, so that would require another core change, similar to the Y Offset one.

Currently, [x advancement is determined](https://github.com/AdamsLair/duality/blob/master/Duality/Resources/Font.cs#L901-L904) by a glyph's `Width - OffsetX`, which would need to be replaced with a new `AdvanceX` field.

**Comment by ilexp** [comment]
Ah, by the way: Your importer currently creates a Pixmap output as well as a Font output. It doesn't actually create it itself as far as I can see, but it does specify that it produces a Pixmap Resource as an output. This is not required! Instead of adding a predicted Pixmap output [here](https://github.com/mika76/duality-bitmapfont-importer/blob/master/DualityBitmapImporter/DualityBitmapImporter/BitmapFontImpoter.cs#L158), you should instead specify that you handled the pixmap's image source file! This will tell the system that it shouldn't ask other importers to handle the file, because you already did.

**Comment by mika76** [comment]
> When looking at a BMFont file (especially in one with outlines and shadow effects), you can see that in almost all cases, xadvance is different from width, so I think that this needs to be addressed. Unfortunately, Duality Fonts do not distinguish between width and x-advancement, so that would require another core change, similar to the Y Offset one.

Cool no problem, I can add that easily

> Ah, by the way: Your importer currently creates a Pixmap output as well as a Font output. It doesn't actually create it itself as far as I can see, but it does specify that it produces a Pixmap Resource as an output. This is not required! Instead of adding a predicted Pixmap output here, you should instead specify that you handled the pixmap's image source file. This will tell the system that it shouldn't ask other importers to handle the file, because you already did.

I actually handled that purely so that the pixmap file get's copied to the Media folder. I'm not sure it's required, but it _is_ part of the source so I thought it was useful. I'll play with what you said.

**Comment by mika76** [comment]
> I have implemented a better readonly-switch in the FontPropertyEditor, so you can now switch kerning and various other properties on and off without an embedded .ttf Font. Let me know if that fixes the issue for your scenario.

That's brilliant! I tried a bunch of things last night, but that property editor code is some complex stuff :laughing:  I couldn't figure out how to only disable certain properties.

**Comment by ilexp** [comment]
> That's brilliant! I tried a bunch of things last night, but that property editor code is some complex stuff :laughing:

Yep, not the most convenient and simple part of Duality :smile: Figured it'd be helpful if I just quickly looked into this.

> I actually handled that purely so that the pixmap file get's copied to the Media folder. I'm not sure it's required, but it is part of the source so I thought it was useful. I'll play with what you said.

Completely agree! As long as you tell the system you're handling the input file, it should actually get copied, so that should be fine. 

(I suspect the reason it is copied right now is because the system thinks no-one has handled the .png file, then forwards it to the Pixmap importer, which will take care of it. This should not be relied on though.)

> Cool no problem, I can add that easily

Great! I'll be waiting for another Pull Request then. Let me know if you need anything.

**Comment by ilexp** [comment]
Just merged my recent progress on #321 into the `master` branch. It is now possible to specify custom import parameters using `IAssetImportEnvironment.GetOrInitParameter<T>`. These parameters then become user-editable in the editor. Might be useful in order to make a BMFont import configurable, if there are options that would make sense.

**Comment by mika76** [comment]
Will check it out thanks. A bit sick at the moment but will get back into thigs soon :smile:

**Comment by ilexp** [comment]
Completed issue #322, which means all Font import data has been moved to the new import parameters API in the new custom AssetInfo data attachments. It's backwards compatible for regular editor usage, but you will likely have to update your BMFont importer to match the new `SetGlyphData` method signature that was introduced. 

There is also no longer any dependency on an `EmbeddedTrueTypeFont` property and it only exists for binary compatibility reasons.

Just an FYI so you know what's Font / import code has been up to recently :)

**Comment by mika76** [comment]
Awesome, can't wait to play with it. Will you be updating the nugets?

**Comment by ilexp** [comment]
Definitely, but not yet. This is a big change and it comes with a lot of potential to break stuff. I'd like to keep this code-only for testing until I'm sure I haven't accidentally broken something crucial.

**Comment by ilexp** [comment]
Rediscovered the issue during a general issue cleanup and I'd like to wrap this up. From reading some of the old comments, I'd assume that the importer is in a WiP state before the v2.x Font update was released. Meanwhile, there was more work done on the v3.0 development branch, adding support for specifying kerning pairs explicitly in the importer. 

@mika76 So.. it's been a while, but if you are willing to pick up work on this again, we could check and see how it would make sense to proceed here:

- Integrate it into `master` without the new kerning support?
- Integrate it into `develop-3.0` making use of the new kerning support?
- Something else?

Alternatively, we could re-assign this issue to any willing contributor, who could approach it either from scratch, or make use of the work that was done before. My priority here would be to clear up which state this issue is in.

**Comment by mika76** [comment]
Hi @ilexp I'm not sure I have the time for this now - I am quite happy if someone want to take over and continue the work. I think it's a great addition although I'm not sure if anybody's using it or not...

**Comment by ilexp** [comment]
@mika76 Alright, thanks for the quick response, and thank you for the previous work on this :) This is **up for grabs** then. 

For maintenance reasons, let's move this into the core as well and re-approach this issue with the goal of submitting, iterating on and merging a Pull Request.

### ToDo

- Investigate the current state of the plugin, as well as compatibility with the latest v2.x of Duality.
- Introduce the importer into the core / master branch.
- Test whether it works, polish if / as needed and consider adjustments to bring it in line with the latest Duality version, should there be any required.
- Prepare a Pull Request on this, iterate on it through review stages. Expect change requests, since core additions will be checked a lot more in detail than separate plugins.
